### ISSUE TITLE
IAM: Python type IPrincipal is incompatible with ServicePrincipal

### ISSUE THREAD

**Issue by fraghag** [aws/aws-cdk#31454]
### Describe the bug

This bug was already reported previously #30431.

Not sure why it was not addressed and the issue closed, so opening a new one. 
The AWS plugin in VSCode is quite extensive so it seems to be a very encouraged tool to use for AWS development, but this bug makes the coding experience very poor.

Also the `_statement` parameter looks very weird
```
principal = iam.AccountPrincipal("123456789012")
principal.add_to_principal_policy(_statement=iam.PolicyStatement())
```
and if it's called like that instead of `statement` like it is in `IPrincipal` for a reason it's seems like an ugly hack.

If you think that you are doing the correct thing with regards to python typing then we should start a discussion about it with pyright developers.

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

I'm expecting to not get typing errors when writing the code as it is intended.

### Current Behavior

Raises typing error.

### Reproduction Steps

See #30431

### CDK CLI Version

2.154.1 (build febce9d)

### Node.js Version

v20.12.2

### OS

Ubuntu 22.04.4 LTS

### Language

Python

### Language Version

Python 3.12

**Comment by pahud** [comment]
Yes i can produce this error when I install `pyright` VSCode extension

<img width="1059" alt="image" src="https://github.com/user-attachments/assets/f8b64046-0a9f-4705-a9b9-c43e722f2534">


But I won't have that if I install [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance).

I guess this might be a Pyright bug?

**Comment by fraghag** [comment]
Pylance uses pyright under the hood. But pylance does not highlight typing errors by default, you have to explicitly tell it to in the settings with:
```
    "python.analysis.typeCheckingMode": "standard",
```

And also I first checked this with pyright and the guy there says that pyright is doing the right thing and his argument makes sense to me.
https://github.com/microsoft/pyright/issues/8984

**Comment by gshpychka** [comment]
This is not a Pyright bug - in Python, argument names have to match.

This is a limitation of JSII - tracked here: https://github.com/aws/jsii/issues/4541

**Comment by xaviergmail** [comment]
This happens with a lot of other interfaces as well, but not all of them for some reason

**Comment by pacohope** [comment]
I don't know if it matters, but the [sample code provided for `CompositePrincipal`](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_iam/CompositePrincipal.html) causes this same error. So anyone following the examples in the official docs will be confronted by type errors.

Today, that sample code is:
```python
role = iam.Role(self, "RDSDirectoryServicesRole",
    assumed_by=iam.CompositePrincipal(
        iam.ServicePrincipal("rds.amazonaws.com"),
        iam.ServicePrincipal("directoryservice.rds.amazonaws.com")),
    managed_policies=[
        iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AmazonRDSDirectoryServiceAccess")
    ]
)
```

If there is a straightforward workaround, until the root cause is fixed, it would be handy to either add it to the docs or at least add it as a comment here on this issue.

<img width="1958" height="836" alt="Image" src="https://github.com/user-attachments/assets/597fc51c-3c53-423b-ba1d-d1c9f0c8f790" />

**Comment by fraghag** [comment]
Well, it's pretty much like a linting error you cannot fix.

Not sure I would call it a "workaround" since it's not a problem that has an impact on runtime. But if you want to silence the message you can add the `# type: ignore`:
``` python
role = iam.Role(self, "RDSDirectoryServicesRole",
    assumed_by=iam.CompositePrincipal(  # type: ignore
        iam.ServicePrincipal("rds.amazonaws.com"),
        iam.ServicePrincipal("directoryservice.rds.amazonaws.com")),
    managed_policies=[
        iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AmazonRDSDirectoryServiceAccess")
    ]
)
```
